openapi.yaml

```diff
@@ -471,6 +591,173 @@ paths:
+              type: string
+        exposed:
+          type: boolean
+        exposed-endpoints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/ExposedEndpoint'
+        life:
+          type: string
+        principal:
+          type: boolean
+        remote:
+          type: boolean
+        tag:
+          type: string
+      additionalProperties: false
+      required:
+        - tag
+        - principal
+        - exposed
+        - remote
+        - life
+    ApplicationSetCharm:
+      type: object
+      properties:
+        application:
+          type: string
+        channel:
+          type: string
+        charm-origin:
+          $ref: '#/components/schemas/CharmOrigin'
+        charm-url:
+          type: string
+        config-settings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        config-settings-yaml:
+          type: string
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        force:
+          type: boolean
+        force-base:
+          type: boolean
+        force-units:
+          type: boolean
+        generation:
+          type: string
+        resource-ids:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        storage-constraints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/StorageConstraints'
+      additionalProperties: false
+      required:
+        - application
+        - generation
+        - charm-url
+        - channel
+        - force
+        - force-units
+        - force-base
+    Base:
+      type: object
+      properties:
+        channel:
+          type: string
+        name:
+          type: string
+      additionalProperties: false
+      required:
+        - name
+        - channel
+    CharmRelation:
+      type: object
+      properties:
+        interface:
+          type: string
+        limit:
+          type: integer
+        name:
+          type: string
+        optional:
+          type: boolean
+        role:
+          type: string
+        scope:
+          type: string
+      additionalProperties: false
+      required:
+        - name
+        - role
+        - interface
+        - optional
+        - limit
+        - scope
+    CharmOrigin:
+      type: object
+      properties:
+        architecture:
+          type: string
+        base:
+          $ref: '#/components/schemas/Base'
+        branch:
+          type: string
+        hash:
+          type: string
+        id:
+          type: string
+        instance-key:
+          type: string
+        revision:
+          type: integer
+        risk:
+          type: string
+        source:
+          type: string
+        track:
+          type: string
+        type:
+          type: string
```

Comment:
What does this denote? `principal` or `subordinate` or something else?

Reply:
That's if it's for a `charm`  or a ` bundle` . I'll make it a simple enum.

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/charmhub/transport/common.go#L21

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/cmd/juju/application/store/charmadapter.go#L88C2-L88C2